File: Diagrammatize.Design.Pattern/code/cxx/ChainOfResponsibility/Sample/Support.cpp
#include "Support.hpp"
#include <iostream>

Support::Support(const std::string& name_): name(name_), next(nullptr) {
}

Support* Support::setNext(Support* next_) {
    next = next_;
    return next;
}

void Support::support(Trouble* trouble) {
    if (resolve(trouble)) {
        done(trouble);
    } else if (next != nullptr) {
        next->support(trouble);
    } else {
        fail(trouble);
    }
}

std::string Support::toString() const {
    return "[" + name + "]";
}

void Support::done(Trouble* trouble) {
    std::cout << *trouble << " is resolved by " << *this << "." << std::endl;
}

void Support::fail(Trouble* trouble) {
    std::cout << *trouble << " cannot be resolved." << std::endl;
}

std::ostream& operator<<(std::ostream& os, const Support& s) {
    os << s.toString();
    return os;
}

File: Diagrammatize.Design.Patterns/code/python/ChainOfResponsibility/Sample/Support.py
import abc

class Support(abc.ABC):
    def __init__(self, name):
        self.name = name
        self.next = None

    def setNext(self, next_):
        self.next = next_
        return next_

    def support(self, trouble):
        if self.resolve(trouble):
            self.done(trouble)
        elif self.next is not None:
            self.next.support(trouble)
        else:
            self.fail(trouble)

    def __str__(self):
        return "[{}]".format(self.name)

    @abc.abstractmethod
    def resolve(self, trouble):
        pass

    def done(self, trouble):
        print("{} is resolved by {}.".format(trouble, self))

    def fail(self, trouble):
        print("{} cannot be resolved.".format(trouble))
